src/lib/stream/chacha/chacha.cpp

```diff
@@ -173,4 +176,70 @@ std::string ChaCha::name() const
    return "ChaCha(" + std::to_string(m_rounds) + ")";
    }
 
+secure_vector<byte> ChaCha::get_keystream(u64bit offset, size_t length)
```

Comment:
This API doesn't seem right to me. It would be better to add `StreamCipher::seek(u64bit byte_offset)` which moves the counter (for ChaCha, by manipulating m_state[12,13] and m_position to move to the byte offset within the keystream blocks - you already have the core of the logic in this function). Then encryption of all zeros produces the desired result. The same `seek` API could be implemented by CTR or Salsa.

Implementing something like `get_keystream` then becomes just

```
secure_vector<uint8_t> keystream(length);
cipher->seek(offset);
cipher->encipher(keystream);
return keystream;
```

The way you have it here, you can concurrently call `get_keystream` at high offsets and encrypt normally with the from-zero offsets. But that doesn't seem like a practical advantage, and doing it this way requires keeping a second copy of the state just in order to support this API. Having a single (per-object) byte position is simpler and seems to support the use case just as well.